In [1]:

import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from pandas import read_csv
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier



$r = \text{NBR_cas}''(t)$

$AVG\ P(r>0)$ sur 14 J

$AVG\ P(r==0)$ sur 14 J

$AVG\ P(r<0)$ sur 14 J

Output $\rightarrow$ \[-1, 1\]

ECART-type $\rightarrow$ certitude

In [2]:
DATA_PATH = "../../data_processed/usa_top200_final.csv"
df = pd.read_csv(DATA_PATH, index_col="Unnamed: 0")

df.drop(labels=["date", 'GEOID', 'NAME', 'cumulative_confirmed', 'smoothed_cumul', 'smoothed_d1'], axis=1, inplace=True)

df = df.sort_values(['shapeID', 'time'])

In [3]:
df.columns

Index(['time', 'shapeID', 'x0_mean', 'x0_amin', 'x0_amax', 'x1_mean',
       'x1_amin', 'x1_amax', 'x2_mean', 'x2_amin', 'x2_amax', 'x3_mean',
       'x3_amin', 'x3_amax', 'x4_mean', 'x4_amin', 'x4_amax', 'x5_mean',
       'x5_amin', 'x5_amax', 'x6_mean', 'x6_amin', 'x6_amax', 'x7_mean',
       'x7_amin', 'x7_amax', 'x8_mean', 'x8_amin', 'x8_amax', 'x9_mean',
       'x9_amin', 'x9_amax', 'x10_mean', 'x10_amin', 'x10_amax', 'smoothed_d2',
       'mobility_transit_stations', 'mobility_retail_and_recreation',
       'mobility_grocery_and_pharmacy', 'mobility_parks',
       'mobility_residential', 'mobility_workplaces'],
      dtype='object')

In [4]:
ids = df["shapeID"].drop_duplicates().values

N_DAYS = 7
times = df["time"].drop_duplicates().values
times_grouped = [times[n:n+N_DAYS] for n in range(0, len(times), N_DAYS)]
times_grouped.pop() # not len 14

cols = df.drop(['time', 'shapeID', 'smoothed_d2'], axis=1).columns
grouped_cols = []
for i in range(0, N_DAYS):
    for label in cols:
        grouped_cols.append('Day {}: {}'.format(i, label))

X_grouped = pd.DataFrame(columns=grouped_cols)
Y_grouped = pd.DataFrame(columns=['y{}'.format(i) for i in range(0,N_DAYS)])

errors = []
bad_format_error = 0
for id in ids:
    for dates in times_grouped:

        id_query = df.shapeID == id
        date_query = df.time.isin(dates)
        query = df[id_query & date_query].drop(['time', 'shapeID'], axis=1)
        
        x_vals = query.drop('smoothed_d2', axis=1).values.flatten()
        y_vals = query.smoothed_d2.values.flatten()

        if(len(x_vals)) != len(grouped_cols):
            bad_format_error+=1
            errors.append(query)
        else:
            df_x_temp = pd.DataFrame(x_vals.reshape(1,-1), columns=grouped_cols)
            df_y_temp = pd.DataFrame(y_vals.reshape(1,-1), columns=['y{}'.format(i) for i in range(0,N_DAYS)])
            X_grouped = X_grouped.append(df_x_temp, ignore_index=True)
            Y_grouped = Y_grouped.append(df_y_temp, ignore_index=True)
print("SUCCESS with {}/{} FORMAT_ERRORS".format(bad_format_error, len(ids) * len(times_grouped)))
X_grouped

SUCCESS with 1544/6000 FORMAT_ERRORS


...        0.655955        -0.526483   
9           0.971203       -0.123988  ...        0.885830         0.519334   
10         -0.572026       -1.747901  ...        0.592805        -0.942254   
11          1.690619       -1.171301  ...        1.345777        -0.497562   
12          0.504974       -0.063797  ...       -0.239277         0.547953   
13          2.460382        0.713396  ...        0.747899        -1.350151   
14          2.299979        0.456923  ...        2.010909        -0.093162   
15          1.860024        1.069297  ...        1.485695        -0.325806   
16         -1.098141       -1.336850  ...        1.421200         0.025232   
17          1.083557       -0.912017  ...        0.438447        -0.669178   
18         -0.432766       -2.478979  ...        0.257215        -0.279282   
19         -0.595954        0.281402  ...        1.052305        -0.810128   
20          0.425586       -0.508199  ...        1.162673        -0.294323   
21          0.224725       -0.341589  ...        1.921513        -0.051242   
22          0.401717        0.021937  ...        0.297753         0.061956   
23          0.496387       -1.435528  ...        1.057908         0.361695   
24          0.814474       -0.040840  ...        1.772568         0.011461   
25          0.798638        1.325288  ...        2.275787        -0.056297   
26          0.771330        1.512955  ...       -0.136915         0.625401   
27          2.195180       -1.490679  ...        1.675063         0.678731   
28          2.768969        0.322313  ...       -0.572637         0.099883   
29          0.688555       -1.027285  ...       -0.139108        -0.904483   
...              ...             ...  ...             ...              ...   
4426        0.402042        0.201593  ...        0.959313        -1.140515   
4427       -0.635118       -0.791195  ...        1.871829        -0.426139   
4428       -0.554105        1.385443  ...        0.023540        -0.340154   
4429        1.813220        1.428736  ...       -0.137208         0.256453   
4430       -0.258801       -0.665899  ...       -0.081053        -0.793794   
4431       -0.097365        1.204118  ...       -0.856816        -0.317188   
4432       -0.134629        0.962184  ...       -0.648771         0.577092   
4433       -0.903090       -0.683681  ...       -0.060146         0.327639   
4434        2.569945        2.369461  ...       -0.375338        -0.750855   
4435       -0.891875        0.822585  ...       -0.755714         0.390674   
4436        1.824783        1.608194  ...        0.538362        -0.476935   
4437        0.620397        0.226145  ...        0.034873         0.674871   
4438       -1.043247       -2.469837  ...        0.782394        -0.796140   
4439        1.869067       -0.508660  ...        1.371464        -0.690760   
4440        0.050623       -0.271389  ...       -0.514830         1.056081   
4441        2.731588        1.237708  ...        0.743513        -0.831935   
4442        1.080385       -0.131577  ...        1.607286         0.502472   
4443        2.422117        1.057522  ...        1.431186        -0.014723   
4444       -1.177906       -1.287767  ...        1.562201         0.375592   
4445       -0.092674       -0.860288  ...        0.254666        -0.156387   
4446       -0.467226       -1.727056  ...        0.136126         0.153212   
4447        1.362042        0.991215  ...        0.250848        -0.627639   
4448        0.279069       -1.456536  ...        0.865020        -0.051602   
4449        0.558434       -0.630391  ...        1.545355         0.311766   
4450        0.168472       -0.257007  ...        0.161467         0.086719   
4451        0.411482       -1.175337  ...        0.616165         0.858540   
4452        0.902341        0.884588  ...        2.104353         0.281485   
4453        0.322729        1.237595  ...        2.155157         0.293042   
4454        1.472600        1.360608  ...       -0.364915   

In [10]:
N_PCs = 11
avg_cols = ["x{}_mean".format(i) for i in range(0,N_PCs)]
min_cols = ["x{}_amin".format(i) for i in range(0,N_PCs)]
max_cols = ["x{}_amax".format(i) for i in range(0,N_PCs)]
generic_cols = ['time', 'shapeID', 'mobility_transit_stations', 'mobility_retail_and_recreation',
       'mobility_grocery_and_pharmacy', 'mobility_parks',
       'mobility_residential', 'mobility_workplaces']

def get_df_from_cols(dataframe, cols):
    X_new = pd.DataFrame()
    for column in dataframe.columns:
        for label in cols:
            if label in column:
                X_new[column] = dataframe[column]
    return X_new

X_no_cams = get_df_from_cols(X_grouped, generic_cols)
X_min = get_df_from_cols(X_grouped, min_cols+generic_cols)
X_max = get_df_from_cols(X_grouped, max_cols+generic_cols)
X_avg = get_df_from_cols(X_grouped, avg_cols+generic_cols)

In [41]:
from sklearn.ensemble import GradientBoostingRegressor 
from sklearn.ensemble import AdaBoostRegressor 
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression

X_train = X_no_cams.to_numpy()
y_train = Y_grouped.mean().to_numpy()

num_folds = 10
seed = 7
scoring = 'neg_mean_squared_error'

models = []
models.append(('GB', GradientBoostingRegressor()))
models.append(('AB', AdaBoostRegressor()))
models.append(('DT', DecisionTreeRegressor()))
models.append(('KNN', KNeighborsRegressor()))
models.append(('LR', LinearRegression()))

results = []
names = []
for name, model in models:
    kfold = KFold(n_splits=num_folds, random_state=seed)
    cv_results = cross_val_score(model, X_train, y_train, cv=kfold, scoring=scoring)
    results.append(cv_results)
    names.append(name)
    msg = "%s %f %f " % (name, cv_results.mean(), cv_results.std())
    print(msg)

DT -2937.549588 1352.933767 
KNN -1483.987277 1304.749986 
LR -1290.123501 1250.325647 


Results: no difference

In [65]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

TRESHOLD = 10

X_train = X_min.to_numpy()
y_train = (Y_grouped > TRESHOLD).sum(axis=1)

num_folds = 10
seed = 7
scoring = 'accuracy'

models = []
models.append(('GB', GradientBoostingClassifier()))
models.append(('AB', AdaBoostClassifier()))
models.append(('DT', DecisionTreeClassifier()))
models.append(('KNN', KNeighborsClassifier()))
models.append(('LR', LogisticRegression()))

results = []
names = []
for name, model in models:
    kfold = KFold(n_splits=num_folds, random_state=seed)
    cv_results = cross_val_score(model, X_train, y_train, cv=kfold, scoring=scoring)
    results.append(cv_results)
    names.append(name)
    msg = "%s %f %f " % (name, cv_results.mean(), cv_results.std())
    print(msg)

GB 0.455341 0.057830 
AB 0.442550 0.058166 
DT 0.346043 0.033837 
KNN 0.412478 0.039655 
LR 0.447716 0.061701 


MEAN OF WEEK W HIGHER 10 CLASSIFICATION

- GB 0.927965 0.019642 
- AB 0.925721 0.020152 
- DT 0.875234 0.026127 
- KNN 0.924821 0.022831 
- LR 0.927741 0.019161 

(w cams)
- GB 0.927293 0.020238 
- AB 0.923704 0.018885 
- DT 0.875671 0.011897 
- KNN 0.923250 0.023082 
- LR 0.922804 0.018741 



N DAYS HIGHER 10 CLASSIFICATION

- GB 0.450636 0.063706 
- AB 0.430657 0.074770 
- DT 0.319354 0.038397 
- KNN 0.410910 0.041697 
- LR 0.460284 0.067121 

(w cams)
- GB 0.455341 0.057830 
- AB 0.442550 0.058166 
- DT 0.346043 0.033837 
- KNN 0.412478 0.039655 
- LR 0.447716 0.061701 